### PR TITLE
Add Periodic update

### PR DESCRIPTION
**Pull request by Mmduh-483**

### CODE REVIEW

**Comment by Mmduh-483** [comment]
@moshe010 only add the logic for add map

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+	for network, podsInterface := range addMap.Items {
+		glog.Infof("PeriodicUpdate(): network %s", network)
+		pods, ok := podsInterface.([]*kapi.Pod)
+		if !ok {
+			glog.Errorf("PeriodicUpdate(): invalid value for add map networks expected pods array \"[]*kubernetes.Pod\", found %T", podsInterface)
+			continue
+		}
+
+		if len(pods) == 0 {
+			return
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: this should be continue as we want to skip to the next network

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+		}
+
+		networkNamespace := pods[0].Namespace
+		netAttInfo, err := d.kubeClient.GetNetworkAttachmentDefinition(networkNamespace, network)
+		if err != nil {
+			glog.Errorf("PeriodicUpdate(): failed to get network attachment %s with error: %v", network, err)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: should be Warning

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+			// skip failed networks
+			continue
+		}
+
+		glog.V(3).Infof("PeriodicUpdate(): network attachment %v", netAttInfo)
+		networkSpec := make(map[string]interface{})
+		err = json.Unmarshal([]byte(netAttInfo.Spec.Config), &networkSpec)
+		if err != nil {
+			glog.Errorf("PeriodicUpdate(): failed to parse network attachment %s with error: %v", network, err)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: change to Warning

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+			// skip failed networks
+			continue
+		}
+		glog.V(3).Infof("PeriodicUpdate(): network attachment spec %+v", networkSpec)
+
+		ibCniSpec, err := utils.GetIbSriovCniSpecFromNetowrkSpec(networkSpec)
+		if err != nil {
+			glog.Errorf("PeriodicUpdate(): failed to parse network attachment spec %+v with error: %v",
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: same as above

File: pkg/daemon/daemon.go [hunk 5]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+				glog.Errorf("PeriodicUpdate(): failed to parse allocated guid %v with error: %v", allocatedGuid,
+					err)
+				continue
+			}
+
+			guidList = append(guidList, guidAddr)
+			podAnnotations := pod.Annotations
+
+			netAnnotations, err := json.Marshal(networks)
+			if err != nil {
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: log warning

File: pkg/daemon/daemon.go [hunk 6]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+				continue
+			}
+
+			ibAnnotation, err := utils.ParseInfiniBandAnnotation(pod)
+			if err != nil {
+				ibAnnotation = map[string]string{network: utils.ConfiguredInfiniBandPod}
+			}
+			ibAnnotationsData, err := json.Marshal(ibAnnotation)
+			if err != nil {
+				continue
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: log warning

File: pkg/daemon/daemon.go [hunk 7]
```diff
@@ -29,11 +44,166 @@ func NewDaemon() (Daemon, error) {
...
+			podAnnotations := pod.Annotations
+
+			netAnnotations, err := json.Marshal(networks)
+			if err != nil {
+				continue
+			}
+
+			ibAnnotation, err := utils.ParseInfiniBandAnnotation(pod)
+			if err != nil {
+				ibAnnotation = map[string]string{network: utils.ConfiguredInfiniBandPod}
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: the ib configured should be after plugin call. and It should be only if we all the plugin.

File: pkg/guid/guid_pool.go [hunk 8]
```diff
@@ -108,9 +109,26 @@ func (p *guidPool) InitPool() error {
 }
 
 /// AllocateGUID allocate guid for the pod
-func (p *guidPool) AllocateGUID() (string, error) {
+func (p *guidPool) AllocateGUID(guid string) (string, error) {
```
Comment (moshe010) [inline pkg/guid/guid_pool.go]: let split it to 2 methods
1. GenerateGUID
2. AllocateGUID

File: pkg/utils/synchronized_map.go [hunk 9]
```diff
@@ -6,7 +6,7 @@ import (
 
 // SynchronizedMap a thread safe string to anything map
 type SynchronizedMap struct {
-	Items map[string]interface{}
+	Items        map[string]interface{}
```
Comment (moshe010) [inline pkg/utils/synchronized_map.go]: why this change?

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -3,14 +3,21 @@ package utils
...
 
+type IbSriovCniSpec struct {
+	Type string `json:"type"`
+	PKey string `json:"pkey"`
+}
+
 const (
 	InfiniBandAnnotation    = "mellanox.infiniband.app"
 	ConfiguredInfiniBandPod = "configured"
+	InfiniBandSriovCNi      = "ib-sriov-cni"
```
Comment (moshe010) [inline pkg/utils/utils.go]: s/InfiniBandSriovCNi/InfiniBandSriovCni

File: pkg/utils/utils.go [hunk 11]
```diff
@@ -55,11 +62,98 @@ func IsPodNetworkConfiguredWithInfiniBand(ibAnnotation map[string]string, networ
...
+		}
+	}
+
+	return nil, fmt.Errorf("cni plugin ib-sriov-cni not found")
+}
+
+func GetPodNetwork(networks []*v1.NetworkSelectionElement, networkName string) (int, *v1.NetworkSelectionElement, error) {
+	for index, network := range networks {
+		if network.Name == networkName {
+			return index, network, nil
```
Comment (moshe010) [inline pkg/utils/utils.go]: why do you need the index?

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: fixed

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: changed

File: pkg/daemon/daemon.go (hunk 4 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: changed

File: pkg/daemon/daemon.go (hunk 5 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: added log

File: pkg/daemon/daemon.go (hunk 6 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: added log

File: pkg/guid/guid_pool.go (hunk 8 above)
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: Splited

File: pkg/utils/synchronized_map.go (hunk 9 above)
Comment (Mmduh-483) [reply pkg/utils/synchronized_map.go]: by mistake, it happens every time when run `make fmt`

File: pkg/utils/utils.go (hunk 10 above)
Comment (Mmduh-483) [reply pkg/utils/utils.go]: changed

File: pkg/utils/utils.go (hunk 11 above)
Comment (Mmduh-483) [reply pkg/utils/utils.go]: After re-checking it is not needed, removed

File: pkg/daemon/daemon.go (hunk 2 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: fixed

File: pkg/daemon/daemon.go (hunk 7 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: fixed

File: pkg/utils/utils.go [hunk 12]
```diff
@@ -55,11 +62,98 @@ func IsPodNetworkConfiguredWithInfiniBand(ibAnnotation map[string]string, networ
...
+	if network.CNIArgs == nil {
+		network.CNIArgs = &map[string]interface{}{}
+	}
+
+	(*network.CNIArgs)["guid"] = guid
+	return nil
+}
+
+// GetIbSriovCniSpecFromNetowrkSpec check if network uses IB-SR-IOV-CNi
+func GetIbSriovCniSpecFromNetowrkSpec(networkSpec map[string]interface{}) (*IbSriovCniSpec, error) {
```
Comment (moshe010) [inline pkg/utils/utils.go]: IsIbSriovCniInNetwork

File: pkg/daemon/daemon.go [hunk 13]
```diff
@@ -29,11 +45,221 @@ func NewDaemon() (Daemon, error) {
...
+				glog.Errorf("PeriodicUpdate(): failed to get pod network spec %s with error: %v",
+					network, err)
+				// skip failed pod
+				continue
+			}
+
+			allocatedGuid, err := utils.GetPodNetworkGuid(podNetwork)
+			if err == nil {
+				// User allocated guid manually
+				if err = d.guidPool.AllocateGUID(allocatedGuid); err != nil {
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: log error

File: pkg/daemon/daemon.go [hunk 14]
```diff
@@ -29,11 +45,221 @@ func NewDaemon() (Daemon, error) {
...
+					continue
+				}
+
+				if err = utils.SetPodNetworkGuid(podNetwork, allocatedGuid); err != nil {
+					glog.Errorf("PeriodicUpdate(): failed to set pod network annotation with error: %v ", err)
+					continue
+				}
+			}
+
+			guidAddr, err := net.ParseMAC(allocatedGuid)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: remove this?

File: pkg/daemon/daemon.go [hunk 15]
```diff
@@ -29,11 +46,306 @@ func NewDaemon() (Daemon, error) {
...
+	go wait.Forever(d.DeletePeriodicUpdate, 2*time.Second)
+	glog.Info("watcher Run():")
 	d.watcher.Run()
 }
+
+func (d *daemon) AddPeriodicUpdate() {
+	glog.Info("AddPeriodicUpdate():")
+	addMap, _ := d.watcher.GetHandler().GetResults()
+	addMap.Lock()
+	time.Sleep(5 * time.Second)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: remove the sleep

File: pkg/guid/guid_pool.go [hunk 16]
```diff
@@ -158,21 +169,29 @@ func (p *guidPool) ReleaseGUID(guid string) error {
 		return err
 	}
 	delete(p.guidPoolMap, guid)
+	delete(p.guidPoolMap, guid)
```
Comment (moshe010) [inline pkg/guid/guid_pool.go]: why duplicate delete?

File: pkg/daemon/daemon.go (hunk 15 above)
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: removed

File: pkg/guid/guid_pool.go (hunk 16 above)
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: Fixed, it should be `delete(p.guidPodNetworkMap, guid)`,
